### DISCUSSION TITLE
Ideas for diagnostic support

### DISCUSSION THREAD

**Discussion by ndonfris** [ndonfris/fish-lsp#37] [Ideas]
Hi all! :wave:

I'm currently working on implementing an updated diagnostic handler [here](https://github.com/ndonfris/fish-lsp/tree/feat/diagnostics-initial). 

___I'm wondering if anyone has any ideas for diagnostic messages to support.___

###### Current Ideas
| name | type | description |
| --- | --- | --- |
| block `end` missing | error | `for\|while\|switch\|function\|if\|begin` is missing an `end` token |
| string missing closing character | error | `'` or `"` is missing from string |
| subshell missing end character | error | `(` needs end character `)` | 
| `command`/`builtin` prefix missing | warning | `command`/`builtin` prefix is missing (user will be able to opt into this behavior through env variables) |
| missing completions file | info | `functions/<file>.fish` does not have completions file |
| autoloaded function file does not have matching function | error | `functions/<file>.fish` does not have a function matching the filename | 
| invalid `fish_lsp_*` env variable value | error | bad value for `fish_lsp_*` variable | 
| single quote includes variable expansion symbol | warning | `echo '$some_variable'` |
| 0 indexed array | error | `echo $argv[0]` - fish indexing starts at 1 |
| unused local symbol | warning | local definition in file is unused |
| missing silent flag | warning | `if command -q some_cmd` or `if set -q some_var` is missing | 
| universal definition in script | warning | don't use `--universal` definitions in fish script, use `--universal` during interactive use (use '--global' instead) |
| alias used instead of function | information | use `function` instead of `alias` |
| function missing return status number | warning | `return 0` missing from function |
| source file not found | error | `source file.fish` file does not exist |
| private function missing underscore prefix | warning | autoloaded function with private function definition is missing `function __private_func` |

**Comment by tippfehlr** [comment]
just curious, why is it recommended to use a function instead of alias?

**Reply by ndonfris** [reply]
Internally, fish-shell defines alias as a wrapper for functions. 

For example you could see this behavior based off of the following two commands:
```fish
alias ls 'exa'
```

```fish
functions -a ls
```
This command will output the following definition for the alias ls: 

```fish
# Defined via `source`
function ls --wraps=exa --description 'alias ls exa'
  exa $argv
        
end
```
Which is the main reasoning I included this diagnostic message. Having said that, I do still use aliases in my own configuration, and just disable the alias diagnostic via:
```fish
set -gx fish_lsp_diagnostic_disable_error_codes 2002
```

**Comment by heygarrett** [comment]
> missing silent flag
> if command -q some_cmd or if set -q some_var is missing

The fish docs for `command` say the `--quiet` flag is deprecated:  
https://fishshell.com/docs/current/cmds/command.html

For `set` the `-q` flag is short for `--query` (same as `command`):  
https://fishshell.com/docs/current/cmds/set.html

I'm currently getting this warning when using `while read`, but `read` doesn't have a `-q` flag at all:  
https://fishshell.com/docs/current/cmds/read.html

**Reply by ndonfris** [reply]
Yeah sorry, I thought I merged this into master earlier. If you want to apply a patch, here is my [current working branch](https://github.com/ndonfris/fish-lsp/tree/feat/parsing-functions-with-event-hook-support) with [some related fixes](https://github.com/ndonfris/fish-lsp/blame/53fcd954f660bfb536c3d36ddafddc6520ff60a4/src/diagnostics/node-types.ts#L219). 

Would the wording change [here](https://github.com/ndonfris/fish-lsp/blob/53fcd954f660bfb536c3d36ddafddc6520ff60a4/src/code-actions/quick-fixes.ts#L172) be preferred? Aside from the `read` command, I'm unsure if the actual behavior of the code-action needs any changes since the `-q` flag is always used in the quickfix. 

I am decently far along on v1.0.10, and was currently planning on including these changes on the upcoming npm release. Would appreciate some testing as edge cases like this are hard to catch :pray:
